onnx/numpy_helper.py

```diff
@@ -4,221 +4,286 @@
 from __future__ import annotations
 
 import sys
+import typing
 from typing import Any, Sequence
 
 import numpy as np
 
-from onnx import MapProto, OptionalProto, SequenceProto, TensorProto, helper, subbyte
-from onnx.external_data_helper import load_external_data_for_tensor, uses_external_data
+import onnx
+import onnx.external_data_helper
+from onnx import helper, subbyte
+
+if typing.TYPE_CHECKING:
+    import numpy.typing as npt
 
 
 def combine_pairs_to_complex(fa: Sequence[int]) -> list[complex]:
+    """Converts alternating [real, imaginary, ...] numbers to complex numbers."""
     return [complex(fa[i * 2], fa[i * 2 + 1]) for i in range(len(fa) // 2)]
 
 
+def _left_shift_16_bits(
+    data: npt.NDArray[np.uint16 | np.int32],
+) -> npt.NDArray[np.uint32]:
+    # The left shifted result is always int64, so we need to convert it back to uint32
+    return (data << 16).astype(np.uint32)
+
+
 def bfloat16_to_float32(
-    data: np.int16 | np.int32 | np.ndarray,
-    dims: int | Sequence[int] | None = None,
-) -> np.ndarray:
-    """Converts ndarray of bf16 (as uint32) to f32 (as uint32).
+    data: npt.NDArray[np.uint16 | np.int32],
+) -> npt.NDArray[np.float32]:
+    """Converts ndarray of bf16 (as uint16 / uint32) to f32.
 
     Args:
         data: A numpy array, empty dimensions are allowed if dims is
             None.
-        dims: If specified, the function reshapes the results.
 
     Returns:
-        A numpy array of float32 with the same dimension if dims is
-        None, or reshaped to dims if specified
+        A numpy array of float32 with the same dimension.
     """
-    shift = lambda x: x << 16  # noqa: E731
-    if dims is None:
-        if len(data.shape) == 0:
-            return shift(np.array([data]).astype(np.int32)).view(np.float32)[0]  # type: ignore[no-any-return]
-        return shift(data.astype(np.int32)).view(np.float32)  # type: ignore[no-any-return]
-    return shift(data.astype(np.int32)).reshape(dims).view(np.float32)  # type: ignore[no-any-return]
-
-
-def _float8e4m3_to_float32_scalar(ival: int, fn: bool, uz: bool) -> np.float32:
```

Comment:
I would keep the code of the old function in the documentation. The logic is easier to read so that the new code can be more easily understood.